src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenUsingDeclarationTests.cs

```diff
@@ -617,6 +628,478 @@ .maxstack  3
 ");
         }
 
+        [Fact]
+        public void InsideTryCatchFinallyBlocksAsync()
+        {
+            string source = @"
+using System;
+using System.Threading.Tasks;
+class C1 : IAsyncDisposable
+{
+    public string Text { get; set; }
+
+    public C1(string text)
+    {
+        Text = text;
+        Console.WriteLine($""Created {Text}"");
+    }
+
+    public ValueTask DisposeAsync()
+    {
+        Console.WriteLine($""Dispose Async {Text}"");
+        return new ValueTask(Task.CompletedTask);
+    }
+}
+class C2
+{
+    public static async Task Main()                                                                                                           
+    {
+        try
+        {
+            await using var x = new C1(""Try"");
+            throw new Exception();
+        }
+        catch
+        {
+            await using var x =  new C1(""Catch"");
+        }
+        finally
+        {
+            await using var x = new C1(""Finally"");
+        }
+    }
+}";
+            string expectedOutput = @"
+Created Try
+Dispose Async Try
+Created Catch
+Dispose Async Catch
+Created Finally
+Dispose Async Finally
+";
+            var compilation = CreateCompilationWithTasksExtensions(source + _asyncDisposable, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(compilation, expectedOutput: expectedOutput).VerifyIL("C2.<Main>d__0.System.Runtime.CompilerServices.IAsyncStateMachine.MoveNext()", @"
```

Comment:
>"C2.<Main>d__0.System.Runtime.CompilerServices.IAsyncStateMachine.MoveNext()" [](start = 83, length = 77)

I've given up on IL verification for large async methods... I usually do a `CompileAndVerify(...).Dump()` verification by hand.